File: 2020 카카오 인턴십/보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        Map<String, Integer> gemList = new HashMap<>();
        for(String gem : gems)
            gemList.put(gem, 1);
        int gemSize = gemList.size();
        int start = 0, dist = 100000;
        
        gemList.clear();
        Queue<String> q = new LinkedList<>();
        int _start = 0;
        for(String gem : gems) {
        	if(gemList.containsKey(gem)) {
                gemList.put(gem, gemList.get(gem) + 1);
            }
            else gemList.put(gem, 1);
            q.add(gem);
            while(true) {
                if(gemList.get(q.peek()) > 1) {
                    gemList.put(q.peek(), gemList.get(q.peek()) - 1);
                    q.remove();
                    _start++;
                }
                else break;
            }
            if(gemList.size() == gemSize && dist > q.size()) {
                start = _start;
                dist = q.size();
            }
        }
        
        int[] answer = { start+1, start + dist };
        return answer;
    }
}

File: 2020 카카오 인턴십/보석 쇼핑.cpp
#include <string>
#include <vector>
#include <map>
#include <queue>

using namespace std;

vector<int> answer = {0,0};
map<string, int> m;
queue<string> q;

vector<int> solution(vector<string> gems) {
    for(string gem : gems)
        m[gem] = 1;
    int gemSize = m.size(), start = 0, len = 100000;
	m.clear();    
    for(int i = 0; i < gems.size(); ++i) {
        m[gems[i]]++;
        q.push(gems[i]);
        while(true) {
            if(m[q.front()] > 1) {
                m[q.front()]--;
                q.pop();
                start++;
            }
            else break;
        }
        if(m.size() == gemSize && len > q.size()) {
            len = q.size();
            answer[0] = start+1;
            answer[1] = start+len;
        }
    }
    return answer;
}
